File: api/_rate_limit.py
"""
Rate limiter for Python serverless functions using Upstash Redis REST API.

Required env vars:
  UPSTASH_REDIS_REST_URL    — Upstash Redis REST URL
  UPSTASH_REDIS_REST_TOKEN  — Upstash REST token

Per-endpoint limit env var (format: "requests,windowSeconds"):
  RATE_LIMIT_MEASURE_LIDAR  — /api/measure_lidar (default: 10,60)

In production (VERCEL_ENV=production):
  Missing config → fail closed.
In development:
  Missing config → warn + allow.
"""
from __future__ import annotations

import hashlib
import json
import os
import sys
import time
import urllib.request
import urllib.error
from typing import Optional

UPSTASH_URL = os.environ.get("UPSTASH_REDIS_REST_URL", "")
UPSTASH_TOKEN = os.environ.get("UPSTASH_REDIS_REST_TOKEN", "")
IS_PROD = os.environ.get("VERCEL_ENV") == "production" or os.environ.get("NODE_ENV") == "production"

_dev_warned = False
_dev_buckets: dict[str, dict] = {}


def _parse_limit(env_key: str, default_reqs: int, default_sec: int) -> tuple[int, int]:
    raw = os.environ.get(env_key, f"{default_reqs},{default_sec}")
    parts = raw.split(",")
    try:
        r = int(parts[0])
        s = int(parts[1]) if len(parts) > 1 else default_sec
    except (ValueError, IndexError):
        return default_reqs, default_sec
    return (r if r > 0 else default_reqs), (s if s > 0 else default_sec)


LIMITS: dict[str, tuple[int, int]] = {
    "measure_lidar": _parse_limit("RATE_LIMIT_MEASURE_LIDAR", 10, 60),
}


def _hash_ip(ip: str) -> str:
    return hashlib.sha256(ip.encode()).hexdigest()[:20]


def _extract_ip(handler) -> str:
    xff = handler.headers.get("X-Forwarded-For", "")
    raw = xff.split(",")[0].strip() if xff else (
        getattr(handler.client_address, "__iter__", None) and handler.client_address[0] or "0.0.0.0"
    )
    return _hash_ip(raw)


def _dev_check(key: str, max_reqs: int, window_sec: int) -> dict:
    now = time.time()
    entry = _dev_buckets.get(key, {"count": 0, "reset_at": now + window_sec})
    if now > entry["reset_at"]:
        entry = {"count": 0, "reset_at": now + window_sec}
    entry["count"] += 1
    _dev_buckets[key] = entry
    remaining = max(0, max_reqs - entry["count"])
    retry_after = max(0, int(entry["reset_at"] - now))
    return {"allowed": entry["count"] <= max_reqs, "remaining": remaining, "retry_after": retry_after}


def _upstash_check(key: str, max_reqs: int, window_sec: int) -> dict:
    now_ms = int(time.time() * 1000)
    window_start_ms = now_ms - window_sec * 1000
    pipeline = [
        ["ZADD", key, "NX", str(now_ms), str(now_ms)],
        ["ZREMRANGEBYSCORE", key, "0", str(window_start_ms)],
        ["ZCARD", key],
        ["PEXPIRE", key, str(window_sec * 1000)],
    ]
    req = urllib.request.Request(
        f"{UPSTASH_URL}/pipeline",
        data=json.dumps(pipeline).encode(),
        headers={
            "Authorization": f"Bearer {UPSTASH_TOKEN}",
            "Content-Type": "application/json",
        },
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=3) as resp:
        data = json.loads(resp.read())
    card_result = data[2]
    count = card_result.get("result", max_reqs + 1) if isinstance(card_result, dict) else (
        card_result if isinstance(card_result, int) else max_reqs + 1
    )
    remaining = max(0, max_reqs - count)
    return {"allowed": count <= max_reqs, "remaining": remaining, "retry_after": window_sec}


def check_rate_limit(handler, endpoint: str) -> dict:
    """Check rate limit. Returns dict with allowed, remaining, retry_after, config_missing."""
    global _dev_warned
    limit = LIMITS.get(endpoint)
    if not limit:
        raise ValueError(f"Unknown rate-limit endpoint: {endpoint}")
    max_reqs, window_sec = limit
    ip_key = _extract_ip(handler)
    key = f"rl:titan:{endpoint}:{ip_key}"

    if not UPSTASH_URL or not UPSTASH_TOKEN:
        if IS_PROD:
            print(f"[rate-limit] UPSTASH not configured in production — failing closed ({endpoint})", file=sys.stderr)
            return {"allowed": False, "remaining": 0, "retry_after": 60, "config_missing": True}
        if not _dev_warned:
            _dev_warned = True
            print("[rate-limit] DEV: Upstash not configured — using in-memory fallback", file=sys.stderr)
        return _dev_check(key, max_reqs, window_sec)

    try:
        return _upstash_check(key, max_reqs, window_sec)
    except Exception as e:
        print(f"[rate-limit] Upstash error: {e}", file=sys.stderr)
        if IS_PROD:
            return {"allowed": False, "remaining": 0, "retry_after": 30}
        return {"allowed": True, "remaining": max_reqs, "retry_after": 0}


def apply_rate_limit(handler, endpoint: str) -> bool:
    """
    Apply rate limit. Sets headers and sends 429 if limited.
    Returns True if request was rate-limited (caller should return immediately).
    """
    limit = LIMITS.get(endpoint)
    result = check_rate_limit(handler, endpoint)
    max_reqs = limit[0] if limit else 0

    handler.send_header("X-RateLimit-Limit", str(max_reqs))
    handler.send_header("X-RateLimit-Remaining", str(result["remaining"]))

    if not result["allowed"]:
        retry = result["retry_after"]
        handler.send_header("Retry-After", str(retry))
        msg = "Service temporarily unavailable" if result.get("config_missing") else "Too many requests"
        body = json.dumps({"error": msg, "retryAfter": retry}).encode()
        handler.send_response(429)
        handler.send_header("Content-Type", "application/json")
        handler.send_header("Content-Length", len(body))
        handler.end_headers()
        handler.wfile.write(body)
        return True
    return False

File: api/_rate-limit.js
'use strict';
/**
 * Rate limiter using Upstash Redis REST API (no npm packages — pure fetch).
 *
 * Required env vars:
 *   UPSTASH_REDIS_REST_URL    — Set to your Upstash Redis REST URL.
 *   UPSTASH_REDIS_REST_TOKEN  — Set to your Upstash REST token.
 *
 * Per-endpoint limits (comma-separated "requests,windowSeconds"):
 *   RATE_LIMIT_MEASURE      — /api/measure    (default: 10,60)
 *   RATE_LIMIT_CONTACT      — /api/contact    (default: 5,600)
 *   RATE_LIMIT_ADMIN_AUTH   — /api/admin-auth (default: 5,900)
 *
 * In production (VERCEL_ENV=production or NODE_ENV=production):
 *   - Missing Upstash config → fail closed (429/503).
 *   - Upstash unreachable → fail closed.
 * In development:
 *   - Missing Upstash config → warn + use in-memory fallback (not durable).
 *   - Upstash unreachable → warn + allow.
 */

const crypto = require('crypto');

const UPSTASH_URL = process.env.UPSTASH_REDIS_REST_URL;
const UPSTASH_TOKEN = process.env.UPSTASH_REDIS_REST_TOKEN;
const IS_PROD =
  process.env.VERCEL_ENV === 'production' ||
  process.env.NODE_ENV === 'production';

function parseLimit(envKey, defaultReqs, defaultSec) {
  const raw = process.env[envKey] || `${defaultReqs},${defaultSec}`;
  const parts = raw.split(',').map(Number);
  return {
    maxReqs: Number.isFinite(parts[0]) && parts[0] > 0 ? parts[0] : defaultReqs,
    windowSec: Number.isFinite(parts[1]) && parts[1] > 0 ? parts[1] : defaultSec,
  };
}

const LIMITS = {
  measure:    parseLimit('RATE_LIMIT_MEASURE',   10,  60),
  contact:    parseLimit('RATE_LIMIT_CONTACT',    5, 600),
  admin_auth: parseLimit('RATE_LIMIT_ADMIN_AUTH', 5, 900),
};

// In-memory dev fallback — not shared across serverless instances.
const _devBuckets = new Map();
let _devWarnedOnce = false;

function _devCheck(key, limit) {
  const now = Date.now();
  const windowMs = limit.windowSec * 1000;
  let entry = _devBuckets.get(key);
  if (!entry || now > entry.resetAt) {
    entry = { count: 0, resetAt: now + windowMs };
  }
  entry.count += 1;
  _devBuckets.set(key, entry);
  const remaining = Math.max(0, limit.maxReqs - entry.count);
  const retryAfter = Math.ceil((entry.resetAt - now) / 1000);
  return { allowed: entry.count <= limit.maxReqs, remaining, retryAfter };
}

async function _upstashCheck(key, limit) {
  const now = Date.now();
  const windowStart = now - limit.windowSec * 1000;
  // Sliding window via pipeline: add timestamp member, prune old, count, expire.
  const pipeline = [
    ['ZADD', key, 'NX', String(now), String(now)],
    ['ZREMRANGEBYSCORE', key, '0', String(windowStart)],
    ['ZCARD', key],
    ['PEXPIRE', key, String(limit.windowSec * 1000)],
  ];
  const res = await fetch(`${UPSTASH_URL}/pipeline`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${UPSTASH_TOKEN}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(pipeline),
    signal: AbortSignal.timeout(3000),
  });
  if (!res.ok) throw new Error(`Upstash HTTP ${res.status}`);
  const data = await res.json();
  // Pipeline result index 2 = ZCARD response
  const cardResult = data[2];
  const count =
    typeof cardResult === 'object' && cardResult !== null
      ? cardResult.result ?? limit.maxReqs + 1
      : typeof cardResult === 'number'
      ? cardResult
      : limit.maxReqs + 1;
  const remaining = Math.max(0, limit.maxReqs - count);
  return { allowed: count <= limit.maxReqs, remaining, retryAfter: limit.windowSec };
}

/**
 * Hash the client IP so it is never stored or logged in plaintext.
 * Uses the first IP from X-Forwarded-For (Vercel is the trusted terminating proxy).
 */
function _hashedIp(req) {
  const xff = req.headers['x-forwarded-for'];
  const raw =
    (typeof xff === 'string' ? xff.split(',')[0].trim() : null) ||
    req.socket?.remoteAddress ||
    '0.0.0.0';
  return crypto.createHash('sha256').update(raw.trim()).digest('hex').slice(0, 20);
}

/**
 * Check the rate limit for the named endpoint.
 *
 * @param {import('http').IncomingMessage} req
 * @param {'measure'|'contact'|'admin_auth'} endpoint
 * @returns {Promise<{allowed: boolean, remaining: number, retryAfter: number, configMissing?: boolean}>}
 */
async function checkRateLimit(req, endpoint) {
  const limit = LIMITS[endpoint];
  if (!limit) throw new Error(`Unknown rate-limit endpoint: ${endpoint}`);

  const ip = _hashedIp(req);
  const key = `rl:titan:${endpoint}:${ip}`;

  if (!UPSTASH_URL || !UPSTASH_TOKEN) {
    if (IS_PROD) {
      console.error(
        `[rate-limit] UPSTASH_REDIS_REST_URL/TOKEN not set — failing closed in production (endpoint: ${endpoint})`
      );
      return { allowed: false, remaining: 0, retryAfter: 60, configMissing: true };
    }
    if (!_devWarnedOnce) {
      _devWarnedOnce = true;
      console.warn(
        '[rate-limit] DEV: Upstash not configured — using in-memory fallback. ' +
        'Set UPSTASH_REDIS_REST_URL + UPSTASH_REDIS_REST_TOKEN for production-accurate behaviour.'
      );
    }
    return _devCheck(key, limit);
  }

  try {
    return await _upstashCheck(key, limit);
  } catch (err) {
    console.error(`[rate-limit] Upstash error for ${endpoint}:`, err.message);
    if (IS_PROD) return { allowed: false, remaining: 0, retryAfter: 30 };
    // Dev: allow through if Upstash unreachable
    return { allowed: true, remaining: limit.maxReqs, retryAfter: 0 };
  }
}

/**
 * Express-style helper. Returns true if the request was rate-limited and a
 * 429 response was already sent. Caller should return immediately in that case.
 *
 * @param {import('http').IncomingMessage} req
 * @param {import('http').ServerResponse} res
 * @param {'measure'|'contact'|'admin_auth'} endpoint
 * @returns {Promise<boolean>}
 */
async function applyRateLimit(req, res, endpoint) {
  const limit = LIMITS[endpoint];
  let result;
  try {
    result = await checkRateLimit(req, endpoint);
  } catch (err) {
    console.error('[rate-limit] checkRateLimit threw:', err.message);
    result = IS_PROD
      ? { allowed: false, remaining: 0, retryAfter: 60 }
      : { allowed: true, remaining: limit?.maxReqs ?? 0, retryAfter: 0 };
  }

  res.setHeader('X-RateLimit-Limit', limit?.maxReqs ?? 0);
  res.setHeader('X-RateLimit-Remaining', result.remaining);

  if (!result.allowed) {
    res.setHeader('Retry-After', result.retryAfter);
    const msg = result.configMissing
      ? 'Service temporarily unavailable'
      : 'Too many requests';
    res.status(429).json({ error: msg, retryAfter: result.retryAfter });
    return true;
  }
  return false;
}

module.exports = { checkRateLimit, applyRateLimit, LIMITS };
